### DISCUSSION TITLE
SetValue and ShouldValidate is not working for textarea

### DISCUSSION THREAD

**Discussion by anjana543** [react-hook-form/react-hook-form#3282] [Q&A] [answered]
SetValue is not working for textarea. I wanna set ShouldValidate to true on value change. Please find my code below,

 useEffect(() => {
    setValue(name, value, { shouldValidate: true, shouldDirty: true });
    setSelectedValue(value);
  }, [value]);

<textarea
          className={(getValues && getValues(name)) || value ? 'filled' : ''}
          id={id}
          rows={rows}
          autocomplete="autoComplete_off"
          readOnly={readOnly}
          onChange={(e) => {
            setValue(name, e.target.value, { shouldValidate: true });
            return e.target.value;
          }}
          defaultValue={selectedValue}
          onChange={(e) => setValue(name, e.target.value, { shouldValidate: true })}
        /> 

I also tried with controller but the issue still exists. 
<Controller
              name={name}
              control={control}
              type="textarea"
              as={<textarea defaultValue={selectedValue} />}
              defaultValue={selectedValue}
              rules={{ required: requiredData?.fields?.value?.value || false }}
            />
Please help me with this query as soon as possible..

Thank you.

**Comment by bluebill1049** [comment, marked as answer]
Could you please share a codesandbox?

**Reply by anjana543** [reply]
Hi @bluebill1049,

Here is the codesandbox link: 
https://codesandbox.io/s/lucid-darkness-e9zym

We have tried all possible ways to set a value dynamically to textarea but it works only with defaultValue option.
Please let me know if you require any further information. This issue is blocking us from releasing our code to production. 

Thanks.

**Reply by bluebill1049** [reply]
why didn't use `<textArea ref={register} />`

`setValue('name', 'whatever')`?

**Reply by anjana543** [reply]
Thank you so much 👍

**Comment by abhinaveksc** [comment]
I am getting this error while adding  ref={register} 

TypeError: path.split is not a function

**Reply by bluebill1049** [reply]
which version are you using? can you reproduce in a codesandbox? https://react-hook-form.com/api/useform/register 

`ref={register}` is v6 syntax, if you are using v7 then it should use spread operator `{...register('name')}`

**Comment by abhinaveksc** [comment]
https://codesandbox.io/s/lucid-darkness-e9zym

here is it.

**Reply by Moshyfawn** [reply]
You're using the latest RHF version, which is `7.25.3`.
```js
"dependencies": {
  "react-hook-form": "latest"
},
```
The way to register your field with V7 is to do
```js
<input {...register("name")} />
```
instead of
```js
<input ref={register} />
```

**Reply by Moshyfawn** [reply]
To use `<Controller />` with your input in V7 you want to do
```js
<Controller
  control={control}
  name="description"
  render={({ field }) => (
    <textarea
      ref={field.ref}
      value={field.value}
      onChange={field.onChange}
      onBlur={field.onBlur}
    />
  )}
/>
```
or simply
```js
<Controller
  control={control}
  name="description"
  render={({ field }) => (
    <textarea {...field} />
  )}
/>
```

Please, refer to [the V7 RHF docs](https://react-hook-form.com/api/useform/register) to learn about the new API

**Reply by Moshyfawn** [reply]
[CSB example](https://codesandbox.io/s/unruffled-christian-wq6su?file=/src/index.js)

**Comment by abhinaveksc** [comment]
```
  useEffect(() => {
    dispatch(getUserDetails());
    dispatch(getCountries());
    // if(userLoginDetails.length > 0)
    // {
      console.log("userLoginDetails useEffect",userLoginDetails.first_name)
      register('firstName',{
        required: true
      })
      register('lastName',{
        required: true
      })
      setValue("firstName", userLoginDetails.first_name, {
        shouldValidate: true,
      })
      setValue("lastName", userLoginDetails.last_name, {
        shouldValidate: true,
      })
    // }

  }, []);
```
  
  This is what i am doing. value is setting but validation is not working . any solution ?
